In [ ]:
Margento: NLP Intro

NATURAL LANGUAGE PROCESSING--NLP--A Brief Intro

WHAT IS NLP


Natural language processing (NLP) is the interaction between computers and human language. 
More specifically, natural language processing is the computer understanding, analysis, manipulation, and/or generation of natural language ( according to dictionary.com). In a wider context, NLP is a component of artificial intelligence (AI), a field of compuational linguistics, and can employ various types of approaches, including machine learning (ML) and deep learning.

(Machine learning is an application of artificial intelligence (AI) that provides systems the ability to automatically learn and improve from experience without being explicitly programmed. Machine learning focuses on the development of computer programs that can access data and use it learn for themselves.)


If one zooms in, they will realize that a section (the last part) of the dictionary definition above has recently veered into specific directions of its own, and that it is meanwhile a subject in and of itlsef--NLG--, while NLP focuses mainly on understanding, classifying, and analyzing natural language. But these two main directions have indeed always oevrlapped and most likely always will.


The human-computer interaction that constitutes the scope of NLP enables real-world applications like automatic text summarization, sentiment analysis, topic extraction, named entity recognition, parts-of-speech tagging, relationship extraction, stemming, and more. 


This great variety and complexity arises from the very boundless 'nature of the natural'. For, what is natural language, and how many ways of and purposes for using it are out there. Virtually, an infinity. How could one trade in infinity--and obtain palpable results? 


But before getting to infinity, we have the... digits we can 'count on'. NLP is quintessentially based on the ancient overlapping between speaking and counting, between speech and numbers. E.g. tell & tale (~Zahl) & toll (Germ.), tally & count & recount (Lat.)


NLP METHODS & ALGORITHMS

How do we develop NLP algorithms?

To start with, what is a human's algorithm for processing human language?

Humans 'automatically' break down sentences into units of meaning. They parse blocks of texts for syntactic and semantic units.

So we have to first explicitly show the computer how to do this, in a process called tokenization.

After tokenization, we can convert the tokens into a matrix (which is called the bag of words model).

Once we have a matrix, we can develop machine learning algorithm to train a model and predict scores.


MAIN CHARACTERISTICS AND GOALS OF AN NLP/ML ALGORITHM

So what will the algorithm be doing?

The algorithm is going to be taking in a lot of numerical values, along with the scores that are associated with these values. These values (or variables or dimensions) are called features.

We will extract the numerical values from the text.

For example, the number of times the word "computer" appears in a piece of text can be defined as a feature.

The values (features) are derived from the input text, and aggreggated on a per-text basis into feature vectors.

Multiple feature vectors are generally placed together into a feature matrix, where each row represents a piece of text (a document).

The algorithm will discover which of these features is relevant, and which are not.

The relevance is determined by whether or not the features differentiate a high scoring essay/article/poem from a low scoring one.

Therefore, we will want to feed the algorithm features that are specific (they measure as few things as possible) and relevant.


LET US BRIEFLY EXPLAIN THE BASIC TERMS



TOKENIZATION

Given a character sequence and a defined document unit, tokenization is the task of chopping it up into pieces, called tokens. While doing so, soemtimes the task also involves throwing away certain characters, such as punctuation.

Tokens are frequently equated with words.

We need to distinguish between 3 concepts here:
    - TOKENS
    - TYPES
    - TERMS

A TOKEN is an instance of a sequence of characters in some particular document that are grouped together as a useful semantic unit for processing. A TYPE is the class of all tokens containing the same character sequence. A TERM is a (perhaps normalized) type that is included in the IR system's dictionary. (Source: Manning et al. 2008. Introduction to Information Retrieval)


LET US CONSIDER AN EXAMPLE:
    
    "To die, to sleep—
No more—and by a sleep to say we end"

Let's count:

How many tokens: 
14?
How many types:
11?   
How many terms:
4?


BAG OF WORDS

Also known as the vector space model. In this model, a text (such as a sentence or a document) is represented as the bag (multiset) of its words, disregarding grammar and even word order but keeping multiplicity. 

We track down the terms, and then count the tokens. 
Or, in plain EN, the algorithm will count how many times a word appears in a document. Those word counts allow us to compare documents and gauge their similarities for applications like search, document classification, and topic modeling.

The idea is to analyse and classify different “bags of words." And by matching the different categories, we identify which “bag” a certain block of text (test data) comes from. The model represents text as a multiset of words.

In mathematics, a multiset (aka bag or mset) is a modification of the concept of a set that, unlike a set, allows for multiple instances for each of its elements. The positive integer number of instances, given for each element is called the multiplicity of this element in the multiset.

Then we construct an n X t document-term matrix, where n is the number of documents, and t is the number of unique terms.
Each column represents a unique term, and each cell i,j represents how many of term j are in document i.
This is called a simple term frequency bag of words. A more refined apprach to the number of occurrences of terms in a document within a corpus is TFIDF (term frequency--inverse document frequency) of which we will speak in a bit.

The NLP features in this model are the terms.


In Python we have a very efficient machine learning library, scikit-learn, deployable and very effective for NLP tasks.


SCIKIT-LEARN BoW & VECTORIZER USAGE


We need to import a number of useful packages in Python

In [1]:

import sklearn


In [2]:

from sklearn.feature_extraction.text import CountVectorizer

Let's define a vectorizer 

In [3]:
vectorizer = CountVectorizer()



The documents we want to analyze make up a corpus. We need to have that corpus made available to the machine.


Let us suppose we have a corpus of 4 short documents:

In [4]:

corpus = ["I am interested in NLP and would like to learn more about vectors and vectorizers for language processing and NLP in general.",

"For those who like to experiment with vectors in dealing with various data, it might be useful to work with NLP vectorizers.",

"A significant computational experiment that would involve language will also have to involve vectors, so you will want to look into vectorizers.",

"Even if you are not interested in NLP, but you are dealing with data, you can use vectors to organize your data."]

In [5]:

X = vectorizer.fit_transform(corpus)

#fit_transform both determines which tokens it will count, and how they correspond to entries in the count vector. 
#It also gives you the count vectors for the data. 

In [6]:
X

<4x50 sparse matrix of type '<class 'numpy.int64'>'
	with 72 stored elements in Compressed Sparse Row format>

In [7]:
print(X)

  (0, 14)	1
  (0, 31)	1
  (0, 22)	1
  (0, 13)	1
  (0, 40)	1
  (0, 41)	1
  (0, 0)	1
  (0, 27)	1
  (0, 23)	1
  (0, 36)	1
  (0, 24)	1
  (0, 47)	1
  (0, 3)	3
  (0, 28)	2
  (0, 17)	2
  (0, 18)	1
  (0, 2)	1
  (1, 46)	1
  (1, 38)	1
  (1, 5)	1
  (1, 26)	1
  (1, 21)	1
  (1, 9)	1
  (1, 39)	1
  (1, 10)	1
  :	:
  (2, 32)	1
  (2, 12)	1
  (2, 22)	1
  (2, 40)	1
  (2, 41)	1
  (2, 36)	2
  (2, 47)	1
  (3, 49)	1
  (3, 30)	1
  (3, 37)	1
  (3, 7)	1
  (3, 6)	1
  (3, 29)	1
  (3, 4)	2
  (3, 16)	1
  (3, 11)	1
  (3, 48)	3
  (3, 9)	2
  (3, 10)	1
  (3, 45)	1
  (3, 41)	1
  (3, 36)	1
  (3, 28)	1
  (3, 17)	1
  (3, 18)	1


The in-built tokenizer extracts words of at least two characters (but the parameters can be adjusted by the user). 
We can call that function and use it specifically:

In [8]:
analyze = vectorizer.build_analyzer()

In [9]:
analyze("I am interested in NLP and would like to learn more about vectors and vectorizers for language processing.")

['am',
 'interested',
 'in',
 'nlp',
 'and',
 'would',
 'like',
 'to',
 'learn',
 'more',
 'about',
 'vectors',
 'and',
 'vectorizers',
 'for',
 'language',
 'processing']

In [10]:
list0 = list(analyze("I am interested in NLP and would like to learn more about vectors and vectorizers for language processing."))

In [11]:
len(list0)

17


Remember in the Shakesperean quote above we reduced the number of terms significantly by counting some of them out? Which ones? 
The stopwords, right?

How do we get the machine to remove stopwords? 
We need a list of them and we need to tell the machine those are stopwords.

In [12]:
import nltk

Now we can create the list based off of the nltk corpus:

In [13]:

stop_list = set(nltk.corpus.stopwords.words("english"))

Then we need to redefine the vectorizer so that its configuration takes into account (and therefore counts out) stopwords:

In [14]:

vectorizer = CountVectorizer(stop_words = stop_list)

Then we redefine the analyzer based on this new vectorizer that removes the stopwords:

In [15]:
analyze = vectorizer.build_analyzer()

Let us now run it again on the same document:

In [16]:
analyze("I am interested in NLP and would like to learn more about vectors and vectorizers for language processing.")

['interested',
 'nlp',
 'would',
 'like',
 'learn',
 'vectors',
 'vectorizers',
 'language',
 'processing']

Do we need to count them again (on our... digits) as we earlier did with the Shakesperean lines to see how many terms were removed, or can we automatize this? (EXERCISE)

In [17]:
list1 = analyze("I am interested in NLP and would like to learn more about vectors and vectorizers for language processing.")

In [18]:
print(17 - len(list1))

8


The vectorization of the whole corpus is also going to look different now that we have factored in stopwords:

In [19]:
X = vectorizer.fit_transform(corpus)

In [20]:
X

<4x26 sparse matrix of type '<class 'numpy.int64'>'
	with 40 stored elements in Compressed Sparse Row format>

In [25]:
print(X)

  (0, 6)	1
  (0, 16)	1
  (0, 9)	1
  (0, 21)	1
  (0, 22)	1
  (0, 10)	1
  (0, 11)	1
  (0, 25)	1
  (0, 14)	2
  (0, 7)	1
  (1, 24)	1
  (1, 19)	1
  (1, 13)	1
  (1, 2)	1
  (1, 20)	1
  (1, 3)	1
  (1, 5)	1
  (1, 21)	1
  (1, 22)	1
  (1, 11)	1
  (1, 14)	1
  (2, 12)	1
  (2, 23)	1
  (2, 0)	1
  (2, 8)	2
  (2, 1)	1
  (2, 17)	1
  (2, 5)	1
  (2, 9)	1
  (2, 21)	1
  (2, 22)	1
  (2, 25)	1
  (3, 15)	1
  (3, 18)	1
  (3, 4)	1
  (3, 2)	2
  (3, 3)	1
  (3, 22)	1
  (3, 14)	1
  (3, 7)	1


Although it prints out that way, X is actually a matrix. To better understand its content we can convert it to an array:
    

In [21]:
X.toarray()

array([[0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 2, 0, 1, 0, 0, 0, 0, 1,
        1, 0, 0, 1],
       [0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1,
        1, 0, 1, 0],
       [1, 1, 0, 0, 0, 1, 0, 0, 2, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1,
        1, 1, 0, 1],
       [0, 0, 2, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 0, 0, 0]], dtype=int64)

But what do these numbers represent? Can we trace them back to the initial words and thus retrieve the interpretation of the columns?

In [22]:
vectorizer.get_feature_names() 

['also',
 'computational',
 'data',
 'dealing',
 'even',
 'experiment',
 'general',
 'interested',
 'involve',
 'language',
 'learn',
 'like',
 'look',
 'might',
 'nlp',
 'organize',
 'processing',
 'significant',
 'use',
 'useful',
 'various',
 'vectorizers',
 'vectors',
 'want',
 'work',
 'would']

How can we track automatically which term is which column?

In [23]:
list2 = vectorizer.get_feature_names() 

In [24]:
list2[8]

'involve'

Does this term occur in a certain document? Say the second document?

In [25]:
Y = list(X.toarray())

In [26]:
Y[1][8]

0

In [27]:
Y[0][8]

0

In [28]:
Y[2][8] 

2

EXERCISE 

Please write the code to find out whether the last document contains this term. 


What if we want to get a better sense of what happens in terms of terms not only in individual document but also across the corpus/archive/dataset?
For now we know very accurately how important a term is for every single document. 
But how relevant is it for the whole corpus and more importantly, where does a term's frequency place a document in comparison to to the other documents in the corpus/archive/dataset?


TFIDF


Let us look into a plain intuitive down-to-earth introduction to the concept:
    https://bit.ly/2HZATvi

Let us see now how we can apply this concept in measuring similarities between documents in our corpus.

In [29]:

from sklearn.feature_extraction.text import TfidfTransformer

In [30]:

transformer = TfidfTransformer(smooth_idf=False)
#What does the paramater smooth_idf do? If a word from vocabulary was never seen in the train data, 
#but occures in the test, smooth_idf allows it to be successfully processed. 
#See a relevant example here https://bit.ly/2r7yPYn



In [31]:

tfidf = transformer.fit_transform(Y)

In [32]:
tfidf

<4x26 sparse matrix of type '<class 'numpy.float64'>'
	with 40 stored elements in Compressed Sparse Row format>

In [33]:
print(tfidf)

  (0, 25)	0.2752418429149148
  (0, 22)	0.16256226633758372
  (0, 21)	0.20932851602003819
  (0, 16)	0.3879214194922459
  (0, 14)	0.41865703204007637
  (0, 11)	0.2752418429149148
  (0, 10)	0.3879214194922459
  (0, 9)	0.2752418429149148
  (0, 7)	0.2752418429149148
  (0, 6)	0.3879214194922459
  (1, 24)	0.3842826314790139
  (1, 22)	0.16103739661802227
  (1, 21)	0.20736496861933432
  (1, 20)	0.3842826314790139
  (1, 19)	0.3842826314790139
  (1, 14)	0.20736496861933432
  (1, 13)	0.3842826314790139
  (1, 11)	0.2726600140485181
  (1, 5)	0.2726600140485181
  (1, 3)	0.2726600140485181
  (1, 2)	0.2726600140485181
  (2, 25)	0.21415439885527446
  (2, 23)	0.30182575990136634
  (2, 22)	0.12648303780918257
  (2, 21)	0.16286994025612517
  (2, 17)	0.30182575990136634
  (2, 12)	0.30182575990136634
  (2, 9)	0.21415439885527446
  (2, 8)	0.6036515198027327
  (2, 5)	0.21415439885527446
  (2, 1)	0.30182575990136634
  (2, 0)	0.30182575990136634
  (3, 22)	0.16452841156123668
  (3, 18)	0.39261322075259175
  (3, 1

In [34]:
tfidf.toarray()

array([[0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.38792142, 0.27524184, 0.        , 0.27524184,
        0.38792142, 0.27524184, 0.        , 0.        , 0.41865703,
        0.        , 0.38792142, 0.        , 0.        , 0.        ,
        0.        , 0.20932852, 0.16256227, 0.        , 0.        ,
        0.27524184],
       [0.        , 0.        , 0.27266001, 0.27266001, 0.        ,
        0.27266001, 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.27266001, 0.        , 0.38428263, 0.20736497,
        0.        , 0.        , 0.        , 0.        , 0.38428263,
        0.38428263, 0.20736497, 0.1610374 , 0.        , 0.38428263,
        0.        ],
       [0.30182576, 0.30182576, 0.        , 0.        , 0.        ,
        0.2141544 , 0.        , 0.        , 0.60365152, 0.2141544 ,
        0.        , 0.        , 0.30182576, 0.        , 0.        ,
        0.        , 0.        , 0.30182576, 0.        , 0.        ,
      

EXERCISE 

How can we find out which terms these values correspond to? For instance, for document 1, which term corresponds to the first occurence of the score 0.27266001?

In [ ]:
#your code here

    

Can we get a breakdown of terms and tfidfs for each term in each document?

In [35]:
list3 = list(tfidf.toarray())

In [36]:
for i in range(25):
    if list3[1][i] != 0:
        print(i, list3[1][i])

2 0.2726600140485181
3 0.2726600140485181
5 0.2726600140485181
11 0.2726600140485181
13 0.3842826314790139
14 0.20736496861933432
19 0.3842826314790139
20 0.3842826314790139
21 0.20736496861933432
22 0.16103739661802227
24 0.3842826314790139


Let us say we want to do that for document 0:

In [37]:
for i in range(25):
    print(list2[i], list3[0][i])

also 0.0
computational 0.0
data 0.0
dealing 0.0
even 0.0
experiment 0.0
general 0.3879214194922459
interested 0.2752418429149148
involve 0.0
language 0.2752418429149148
learn 0.3879214194922459
like 0.2752418429149148
look 0.0
might 0.0
nlp 0.41865703204007637
organize 0.0
processing 0.3879214194922459
significant 0.0
use 0.0
useful 0.0
various 0.0
vectorizers 0.20932851602003819
vectors 0.16256226633758372
want 0.0
work 0.0


Can we sort the the items above in increasing/decreasing order?

In [38]:
list4 = []
for i in range(25):
    list4.append((list2[i], list3[0][i]))

In [39]:
sorted(list4, key = lambda t:t[1], reverse = True)

[('nlp', 0.41865703204007637),
 ('general', 0.3879214194922459),
 ('learn', 0.3879214194922459),
 ('processing', 0.3879214194922459),
 ('interested', 0.2752418429149148),
 ('language', 0.2752418429149148),
 ('like', 0.2752418429149148),
 ('vectorizers', 0.20932851602003819),
 ('vectors', 0.16256226633758372),
 ('also', 0.0),
 ('computational', 0.0),
 ('data', 0.0),
 ('dealing', 0.0),
 ('even', 0.0),
 ('experiment', 0.0),
 ('involve', 0.0),
 ('look', 0.0),
 ('might', 0.0),
 ('organize', 0.0),
 ('significant', 0.0),
 ('use', 0.0),
 ('useful', 0.0),
 ('various', 0.0),
 ('want', 0.0),
 ('work', 0.0)]

In [40]:
dt = [('correlation', float)]

In [41]:
import numpy as np

In [42]:
similarity_matrix = np.matrix((tfidf * tfidf.T).A, dtype=dt)

In [43]:
import numpy as np

In [44]:
similarity_matrix = np.matrix((tfidf * tfidf.T).A, dtype=dt)

In [45]:
print(similarity_matrix)

[[(1.        ,) (0.23144825,) (0.17254319,) (0.19211725,)]
 [(0.23144825,) (1.        ,) (0.11253336,) (0.298293  ,)]
 [(0.17254319,) (0.11253336,) (1.        ,) (0.02081005,)]
 [(0.19211725,) (0.298293  ,) (0.02081005,) (1.        ,)]]



How do we sort this list in increasing/decreasing order?


Let us take another example with, this time, larger documents.

In [51]:
corpus1 = ["Mr. Utterson the lawyer was a man of a rugged countenance that was never lighted by a smile; cold, scanty and embarrassed in discourse; backward in sentiment; lean, long, dusty, dreary and yet somehow lovable. At friendly meetings, and when the wine was to his taste, something eminently human beaconed from his eye; something indeed which never found its way into his talk, but which spoke not only in these silent symbols of the after-dinner face, but more often and loudly in the acts of his life. He was austere with himself; drank gin when he was alone, to mortify a taste for vintages; and though he enjoyed the theatre, had not crossed the doors of one for twenty years. But he had an approved tolerance for others; sometimes wondering, almost with envy, at the high pressure of spirits involved in their misdeeds; and in any extremity inclined to help rather than to reprove. I incline to Cain’s heresy, he used to say quaintly: I let my brother go to the devil in his own way. In this character, it was frequently his fortune to be the last reputable acquaintance and the last good influence in the lives of downgoing men. And to such as these, so long as they came about his chambers, he never marked a shade of change in his demeanour. No doubt the feat was easy to Mr. Utterson; for he was undemonstrative at the best, and even his friendship seemed to be founded in a similar catholicity of good-nature. It is the mark of a modest man to accept his friendly circle ready-made from the hands of opportunity; and that was the lawyer’s way. His friends were those of his own blood or those whom he had known the longest; his affections, like ivy, were the growth of time, they implied no aptness in the object. Hence, no doubt the bond that united him to Mr. Richard Enfield, his distant kinsman, the well-known man about town. It was a nut to crack for many, what these two could see in each other, or what subject they could find in common. It was reported by those who encountered them in their Sunday walks, that they said nothing, looked singularly dull and would hail with obvious relief the appearance of a friend. For all that, the two men put the greatest store by these excursions, counted them the chief jewel of each week, and not only set aside occasions of pleasure, but even resisted the calls of business, that they might enjoy them uninterrupted. It chanced on one of these rambles that their way led them down a by-street in a busy quarter of London. The street was small and what is called quiet, but it drove a thriving trade on the weekdays. The inhabitants were all doing well, it seemed and all emulously hoping to do better still, and laying out the surplus of their grains in coquetry; so that the shop fronts stood along that thoroughfare with an air of invitation, like rows of smiling saleswomen. Even on Sunday, when it veiled its more florid charms and lay comparatively empty of passage, the street shone out in contrast to its dingy neighbourhood, like a fire in a forest; and with its freshly painted shutters, well-polished brasses, and general cleanliness and gaiety of note, instantly caught and pleased the eye of the passenger.", "One morning, when Gregor Samsa woke from troubled dreams, he found himself transformed in his bed into a horrible vermin.  He lay on his armour-like back, and if he lifted his head a little he could see his brown belly, slightly domed and divided by arches into stiff sections.  The bedding was hardly able to cover it and seemed ready to slide off any moment.  His many legs, pitifully thin compared with the size of the rest of him, waved about helplessly as he looked. What has happened to me? he thought.  It was not a dream.  His room, a proper human room although a little too small, lay peacefully between its four familiar walls.  A collection of textile samples lay spread out on the table - Samsa was a travelling salesman - and above it there hung a picture that he had recently cut out of an illustrated magazine and housed in a nice, gilded frame.  It showed a lady fitted out with a fur hat and fur boa who sat upright, raising a heavy fur muff that covered the whole of her lower arm towards the viewer. Gregor then turned to look out the window at the dull weather. Drops of rain could be heard hitting the pane, which made him feel quite sad.  How about if I sleep a little bit longer and forget all this nonsense, he thought, but that was something he was unable to do because he was used to sleeping on his right, and in his present state could not get into that position.  However hard he threw himself onto his right, he always rolled back to where he was.  He must have tried it a hundred times, shut his eyes so that he would not have to look at the floundering legs, and only stopped when he began to feel a mild, dull pain there that he had never felt before.", "It is impossible to follow here the varying fortunes of the Italian states, which in 1507 were controlled by France, Spain, and Germany, with results that have lasted to our day; we are concerned with those events, and with the three great actors in them, so far only as they impinge on the personality of Machiavelli. He had several meetings with Louis XII of France, and his estimate of that monarch's character has already been alluded to. Machiavelli has painted Ferdinand of Aragon as the man who accomplished great things under the cloak of religion, but who in reality had no mercy, faith, humanity, or integrity; and who, had he allowed himself to be influenced by such motives, would have been ruined. The Emperor Maximilian was one of the most interesting men of the age, and his character has been drawn by many hands; but Machiavelli, who was an envoy at his court in 1507-8, reveals the secret of his many failures when he describes him as a secretive man, without force of character--ignoring the human agencies necessary to carry his schemes into effect, and never insisting on the fulfilment of his wishes. The remaining years of Machiavelli's official career were filled with events arising out of the League of Cambrai, made in 1508 between the three great European powers already mentioned and the pope, with the object of crushing the Venetian Republic. This result was attained in the battle of Vaila, when Venice lost in one day all that she had won in eight hundred years. Florence had a difficult part to play during these events, complicated as they were by the feud which broke out between the pope and the French, because friendship with France had dictated the entire policy of the Republic. When, in 1511, Julius II finally formed the Holy League against France, and with the assistance of the Swiss drove the French out of Italy, Florence lay at the mercy of the Pope, and had to submit to his terms, one of which was that the Medici should be restored. The return of the Medici to Florence on 1st September 1512, and the consequent fall of the Republic, was the signal for the dismissal of Machiavelli and his friends, and thus put an end to his public career, for, as we have seen, he died without regaining office.", "Having had some time at my disposal when in London, I had visited the British Museum, and made search among the books and maps in the library regarding Transylvania; it had struck me that some foreknowledge of the country could hardly fail to have some importance in dealing with a nobleman of that country. I find that the district he named is in the extreme east of the country, just on the borders of three states, Transylvania, Moldavia and Bukovina, in the midst of the Carpathian mountains; one of the wildest and least known portions of Europe. I was not able to light on any map or work giving the exact locality of the Castle Dracula, as there are no maps of this country as yet to compare with our own Ordnance Survey maps; but I found that Bistritz, the post town named by Count Dracula, is a fairly well-known place. I shall enter here some of my notes, as they may refresh my memory when I talk over my travels with Mina. In the population of Transylvania there are four distinct nationalities: Saxons in the South, and mixed with them the Wallachs, who are the descendants of the Dacians; Magyars in the West, and Szekelys in the East and North. I am going among the latter, who claim to be descended from Attila and the Huns. This may be so, for when the Magyars conquered the country in the eleventh century they found the Huns settled in it. I read that every known superstition in the world is gathered into the horseshoe of the Carpathians, as if it were the centre of some sort of imaginative whirlpool; if so my stay may be very interesting. (Mem., I must ask the Count all about them.) I did not sleep well, though my bed was comfortable enough, for I had all sorts of queer dreams. There was a dog howling all night under my window, which may have had something to do with it; or it may have been the paprika, for I had to drink up all the water in my carafe, and was still thirsty. Towards morning I slept and was wakened by the continuous knocking at my door, so I guess I must have been sleeping soundly then. I had for breakfast more paprika, and a sort of porridge of maize flour which they said was mamaliga, and egg-plant stuffed with forcemeat, a very excellent dish, which they call impletata. (Mem., get recipe for this also.) I had to hurry breakfast, for the train started a little before eight, or rather it ought to have done so, for after rushing to the station at 7:30 I had to sit in the carriage for more than an hour before we began to move. It seems to me that the further east you go the more unpunctual are the trains. What ought they to be in China?"]

In [ ]:
# The first pages of Dr Jekyll and Mr Hyde, Metamorphosis, The Prince, and Dracula

In [52]:
X1 = vectorizer.fit_transform(corpus1)

In [53]:
X1

<4x639 sparse matrix of type '<class 'numpy.int64'>'
	with 713 stored elements in Compressed Sparse Row format>

In [54]:
print(X1)

  (0, 421)	1
  (0, 429)	1
  (0, 80)	1
  (0, 302)	1
  (0, 399)	1
  (0, 240)	1
  (0, 92)	1
  (0, 242)	1
  (0, 57)	1
  (0, 432)	1
  (0, 514)	1
  (0, 416)	1
  (0, 232)	1
  (0, 220)	1
  (0, 210)	1
  (0, 392)	1
  (0, 132)	1
  (0, 106)	1
  (0, 510)	1
  (0, 420)	1
  (0, 169)	1
  (0, 98)	1
  (0, 320)	1
  (0, 87)	1
  (0, 213)	1
  :	:
  (3, 227)	1
  (3, 6)	1
  (3, 261)	1
  (3, 334)	1
  (3, 44)	1
  (3, 153)	1
  (3, 379)	1
  (3, 546)	1
  (3, 337)	1
  (3, 485)	1
  (3, 209)	1
  (3, 109)	1
  (3, 582)	1
  (3, 622)	2
  (3, 578)	1
  (3, 314)	3
  (3, 349)	1
  (3, 248)	1
  (3, 453)	1
  (3, 409)	1
  (3, 572)	1
  (3, 563)	1
  (3, 225)	2
  (3, 530)	1
  (3, 638)	1


In [55]:
Y1 = X1.toarray()

In [56]:
print(Y1)

[[0 0 0 ... 0 1 1]
 [0 0 0 ... 0 0 0]
 [2 1 1 ... 1 2 0]
 [0 0 0 ... 0 0 1]]


In [57]:
list_words = vectorizer.get_feature_names() 

In [58]:
list_words[0]

'1507'

In [59]:
list_words[55]

'books'


EXERCISE
In which of the documents does it occur? And how many times?

In [60]:
#Your code here 
for i in range(4):
    print(Y1[i][55])


0
0
0
1


In [61]:
tfidf1 = transformer.fit_transform(Y1)

In [62]:
tfidf1

<4x639 sparse matrix of type '<class 'numpy.float64'>'
	with 713 stored elements in Compressed Sparse Row format>

In [63]:
tfidf1.toarray()

array([[0.        , 0.        , 0.        , ..., 0.        , 0.04007232,
        0.04007232],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.12593762, 0.06296881, 0.06296881, ..., 0.06296881, 0.0893565 ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.04296216]])

In [64]:
tfidf1_list = list(tfidf1.toarray())

In [65]:
list_document3 = []
for i in range(638):
    list_document3.append((list_words[i], tfidf1_list[3][i]))

In [66]:
print(list_document3)

[('1507', 0.0), ('1508', 0.0), ('1511', 0.0), ('1512', 0.0), ('1st', 0.0), ('30', 0.0605501730294036), ('able', 0.04296215773608023), ('accept', 0.0), ('accomplished', 0.0), ('acquaintance', 0.0), ('actors', 0.0), ('acts', 0.0), ('affections', 0.0), ('age', 0.0), ('agencies', 0.0), ('air', 0.0), ('allowed', 0.0), ('alluded', 0.0), ('almost', 0.0), ('alone', 0.0), ('along', 0.0), ('already', 0.0), ('also', 0.0605501730294036), ('although', 0.0), ('always', 0.0), ('among', 0.1211003460588072), ('appearance', 0.0), ('approved', 0.0), ('aptness', 0.0), ('aragon', 0.0), ('arches', 0.0), ('arising', 0.0), ('arm', 0.0), ('armour', 0.0), ('aside', 0.0), ('ask', 0.0605501730294036), ('assistance', 0.0), ('attained', 0.0), ('attila', 0.0605501730294036), ('austere', 0.0), ('back', 0.0), ('backward', 0.0), ('battle', 0.0), ('beaconed', 0.0), ('bed', 0.04296215773608023), ('bedding', 0.0), ('began', 0.04296215773608023), ('belly', 0.0), ('best', 0.0), ('better', 0.0), ('bistritz', 0.06055017302940

Now it is a significantly longer list indeed. How do we sort it?

In [67]:
sorted(list_document3, key = lambda t:t[1])

[('1507', 0.0),
 ('1508', 0.0),
 ('1511', 0.0),
 ('1512', 0.0),
 ('1st', 0.0),
 ('accept', 0.0),
 ('accomplished', 0.0),
 ('acquaintance', 0.0),
 ('actors', 0.0),
 ('acts', 0.0),
 ('affections', 0.0),
 ('age', 0.0),
 ('agencies', 0.0),
 ('air', 0.0),
 ('allowed', 0.0),
 ('alluded', 0.0),
 ('almost', 0.0),
 ('alone', 0.0),
 ('along', 0.0),
 ('already', 0.0),
 ('although', 0.0),
 ('always', 0.0),
 ('appearance', 0.0),
 ('approved', 0.0),
 ('aptness', 0.0),
 ('aragon', 0.0),
 ('arches', 0.0),
 ('arising', 0.0),
 ('arm', 0.0),
 ('armour', 0.0),
 ('aside', 0.0),
 ('assistance', 0.0),
 ('attained', 0.0),
 ('austere', 0.0),
 ('back', 0.0),
 ('backward', 0.0),
 ('battle', 0.0),
 ('beaconed', 0.0),
 ('bedding', 0.0),
 ('belly', 0.0),
 ('best', 0.0),
 ('better', 0.0),
 ('bit', 0.0),
 ('blood', 0.0),
 ('boa', 0.0),
 ('bond', 0.0),
 ('brasses', 0.0),
 ('broke', 0.0),
 ('brother', 0.0),
 ('brown', 0.0),
 ('business', 0.0),
 ('busy', 0.0),
 ('cain', 0.0),
 ('called', 0.0),
 ('calls', 0.0),
 ('cambra

The one above comes in increasing order. What if we need the opposite?

In [68]:

sorted(list_document3, key = lambda t:t[1], reverse = True)

[('country', 0.302750865147018),
 ('may', 0.302750865147018),
 ('east', 0.1816505190882108),
 ('maps', 0.1816505190882108),
 ('transylvania', 0.1816505190882108),
 ('known', 0.1288864732082407),
 ('among', 0.1211003460588072),
 ('breakfast', 0.1211003460588072),
 ('count', 0.1211003460588072),
 ('dracula', 0.1211003460588072),
 ('huns', 0.1211003460588072),
 ('magyars', 0.1211003460588072),
 ('mem', 0.1211003460588072),
 ('named', 0.1211003460588072),
 ('ought', 0.1211003460588072),
 ('paprika', 0.1211003460588072),
 ('sort', 0.1211003460588072),
 ('must', 0.08592431547216046),
 ('well', 0.08592431547216046),
 ('found', 0.0653476566547517),
 ('30', 0.0605501730294036),
 ('also', 0.0605501730294036),
 ('ask', 0.0605501730294036),
 ('attila', 0.0605501730294036),
 ('bistritz', 0.0605501730294036),
 ('books', 0.0605501730294036),
 ('borders', 0.0605501730294036),
 ('british', 0.0605501730294036),
 ('bukovina', 0.0605501730294036),
 ('call', 0.0605501730294036),
 ('carafe', 0.0605501730294